packages/styled/types/tests.tsx

```diff
@@ -217,3 +217,45 @@ const Input5 = styled.input`
   // $ExpectError
   ;<StyledCompWithoutAs as={Section} />
 }
+
+{
+  type StyledOptsParameterized = StyledOptions<{ a: string; b: number }>
+
+  const styledOpts0: StyledOptions = {
+    label: 'foo',
+    target: 'bar',
+    shouldForwardProp: p => true
+  }
+  const styledOpts1: StyledOptsParameterized = {
+    label: 'foo',
+    target: 'bar',
+    shouldForwardProp: (p: 'a' | 'b') => true
+  }
+
+  const styledOptsBroken0: StyledOptsParameterized = {
+    label: 'foo',
+    target: 'bar',
+    shouldForwardProp: (p: 'c') => true // $ExpectError
+  }
+  const styledOptsBroken1: StyledOptions = {
+    label: 1, // $ExpectError
+    target: null, // $ExpectError
+    shouldForwardProp: (p1, p2) => true // $ExpectError
+  }
+
+  interface Props {
+    foo: boolean
+  }
+  const options: StyledOptions<JSX.IntrinsicElements['div']> = {
+    label: 'foo',
+    target: 'bar',
+    shouldForwardProp: p => true
+  }
+
+  const test1 = styled('div', options)({ color: 'red' })
+  const test2 = styled('div', options)<Props>({ color: 'red' })
+  const test3 = styled(() => <div className="foo" />, options)({ color: 'red' })
+  const test4 = styled(() => <div className="foo" />, options)<Props>({
+    color: 'red'
+  })
```

Comment:
Those are basically the most interesting test cases that have been added as part of this PR - what I mean by that is that this is sort of the expected "integration point" and we need to make sure that this integration point works OK. 

Those added tests are somewhat broad - not allowing me to quickly see how you are going to utilize the introduced changes. Unless on of those 4 tests did not work before your changes? Could you add something more resembling your use case? Writing down in a comment (backed with an example) what has not been possible before this PR and what will become possible with the changes in this PR would make it far easier to review this.

Sorry for the delay - I'll make sure to iterate over this quickly now cause you have already been waiting for so long. I need to understand the full context behind this PR though and I somewhat feel like I'm missing it (it might be here, among all of the comments etc but I don't have the energy to build it up from those older comments, sorry for that - a short summary would be highly appreciated).

Reply:
One main problem that I'm having with this PR is that it has been somewhat repurposed (which is not a bad thing in itself, it just makes it slightly harder for me to navigate it and to build that mental context for it). It has started as allowing the generic to be optional but it ended up adding a new feature~ to it (the `propName` can not be constrained to specific keys)

Reply:
The context: we're building sub-primitive factories to generate a deeply customizable set of components. Not just theming. These sets also differ in internals, their behaviour, and their composition approach. It's something like SDK for UI Kits.

The only purpose of this PR is to achieve strict consistency between `shouldForwardProp` filter and component `Props` where it's needed. We want to control props forwarding and we try to ensure that every primitive combination is correct by the type checking.

The current implementation allows any filter to handle any props. Moreover, this feature was available in earlier versions of `emotion` https://github.com/antongolub/emotion/commit/ad77ed24b4bfe62d6c80f0498cd7e76863e2f28e#diff-3fae48ecece75c5d911928f5c96c7292ac893baae5232e9868e68446de341401R34

![image](https://user-images.githubusercontent.com/5288046/148656363-05330bd0-a07b-4a1d-9e93-b250d5054c81.png)

Reply:
I understand, there would be little chance to bring such functionality as new. But I'm just trying to get back what was lost ))

Reply:
> The current implementation allows any filter to handle any props.

This I can see, somewhat. But a test case showcasing how something was not an error before and how it is an error now would be helpful in assessing how this change works now.

> I understand, there would be little chance to bring such functionality as new.

I'm not opposed to bringing any functionality as new - I just want to understand as much of the context behind the change as possible 😉 

> The current implementation allows any filter to handle any props. Moreover, this feature was available in earlier versions of emotion antongolub@ad77ed2#diff-3fae48ecece75c5d911928f5c96c7292ac893baae5232e9868e68446de341401R34

This kinda points to `FilteringStyledOptions` that is still available as part of the Emotion types - so I'm not exactly sure how this was available before but it is not now.

Reply:
wrong line, my fault:

https://github.com/antongolub/emotion/commit/ad77ed24b4bfe62d6c80f0498cd7e76863e2f28e#diff-3fae48ecece75c5d911928f5c96c7292ac893baae5232e9868e68446de341401R40

![image](https://user-images.githubusercontent.com/5288046/148684253-c3a40b1c-8820-4b4f-b461-8df590c121d7.png)

Reply:
Just a note - I would still appreciate compelling test cases to be added to this PR before proceeding further.

Reply:
Hmm... But this is exactly how we're using this right now. We take some `props`, and some `propsForwarder` filter, wrap them into `StyledOptions<typeof props>`, and use this type to describe `styledOptions` exportable constant.
```js
const propsForwarder = (p: 'foo') => true
const styledOptsParameterized: StyledOptions<{ foo: string }> = {
    shouldForwardProp: propsForwarder
  }

  const styledOptsParameterizedBroken: StyledOptions<{ foo: string }> = {
    shouldForwardProp: (p: 'bar') => true // $ExpectError
  }
```

Reply:
Done if I understood suggestions correctly) About `FilteringStyledOptions` — I'm not sure that it should be changed here. This pull request definitely doesn't break anything and can be safely merged.

Reply:
I think we should attempt to apply the same change to `FilteringStyledOptions` (with the appropriate tests). If this will prove to be harder than expected - then this can be put aside and we'll just follow with this PR. It's just that I find it weird that those types don't align with each other and I bet that I will be baffled about this in the future when looking through code.

You could also try to check the inference results with `$ExpectType` as suggested in point 1.